crates/bevy_ecs/src/world/mod.rs

```diff
@@ -746,15 +739,67 @@ impl World {
         ticks.is_changed(self.last_change_tick(), self.read_change_tick())
     }
 
-    /// Gets a mutable reference to the resource of the given type, if it exists. Otherwise returns
+    /// Gets a reference to the resource of the given type
+    ///
+    /// # Panics
+    ///
+    /// Panics if the resource does not exist.
+    /// Use [`get_resource`](World::get_resource) instead if you want to handle this case.
+    ///
+    /// If you want to instead insert a value if the resource does not exist,
+    /// use [`get_resource_or_insert_with`](World::get_resource_or_insert_with).
+    #[inline]
+    pub fn resource<R: Resource>(&self) -> &R {
+        self.get_resource().unwrap_or_else(|| {
+            panic!(
+                "Requested resource {} does not exist in the `World`. 
+                Did you forget to add it using `app.add_resource` / `app.init_resource`? 
+                Resources are also implicitly added via `app.add_event`,
+                and can be added by plugins.",
+                std::any::type_name::<R>()
+            )
+        })
+    }
```

Comment:
All the new infallible functions should have `#[track_caller]`, as otherwise, the panic message shows the panic location as inside the fn, instead of its call site.

This also requires manually inlinening `unwrap_or_else`, as it isn't `#[track_caller]`.
```suggestion
    #[inline]
    #[track_caller]
    pub fn resource<R: Resource>(&self) -> &R {
        match self.get_resource() {
            Some(x) => x,
            None => {
                panic!(
                    "Requested resource {} does not exist in the `World`. 
                    Did you forget to add it using `app.add_resource` / `app.init_resource`? 
                    Resources are also implicitly added via `app.add_event`,
                    and can be added by plugins.",
                    std::any::type_name::<R>()
                )
            }
        }
    }
```

Reply:
Does track_caller incur runtime overhead? The rust docs on the subject don't call this out explicitly, but given that it only works on "Rust abi" functions (and they mention intrinsics), it feels like it might.

Reply:
Looks like it will automatically pass one pointer as a hidden argument to the function: https://rustc-dev-guide.rust-lang.org/backend/implicit-caller-location.html#generating-code-for-track_caller-callees

Reply:
Theres also worries that it would produce binary size bloat (if done automatically), so the affect on code size is also a consideration:
https://github.com/rust-lang/rust/issues/70579

Reply:
Maybe `#[cfg_attr(debug_assertions, track_caller)]` would work? That way you wouldn't need to pay in binary size for a feature that's likely unnecessary in release builds.

Reply:
ooh i think i like it

Reply:
lets roll with that

Reply:
While there is some runtime overhead with `track_caller`, I think it will be much smaller due to the `#[inline]`, which should allow LLVM to optimize it away.

As for binary size bloat, I am not sure.

PS.: If `#[cfg_attr(debug_assertions, track_caller)]` is to be done, than the three other usages of `#[track_caller]` in bevy should also be changed. (`single()` and `single_mut()` in `Query` and `entity()` in `Commands`)

Reply:
> While there is some runtime overhead with track_caller, I think it will be much smaller due to the #[inline], which should allow LLVM to optimize it away.

To optimize it away, LLVM would need to know that the resource does exist right? Idk if the bevy_ecs storage boundary would allow it to do that.

Reply:
Yes, the panic is still possible, but I was talking about passing on the provided Location that's optimized away.

One should also consider that `track_caller` tracking is already done here in Bevy, as `.unwrap()` also has that attribute.

Reply:
Ahhh if its already that widespread, this is a non-issue. Only enabling this in debug builds doesn't make sense with that in mind. Thanks for the clarity!